qhbmlib/inference/qnn.py

```diff
@@ -121,22 +121,95 @@ def _sampled_expectation(self, initial_states: tf.Tensor,
       unaveraged expectation values of `observable` against each transformed
       initial state.
     """
-    u = self.circuit + observable.circuit_dagger
-    unique_states, idx, _ = utils.unique_bitstrings_with_counts(initial_states)
-    circuits = u(unique_states)
-    num_circuits = tf.shape(circuits)[0]
-    tiled_values = tf.tile(
-        tf.expand_dims(u.symbol_values, 0), [num_circuits, 1])
-    samples = self._sample_layer(
-        circuits,
-        symbol_names=u.symbol_names,
-        symbol_values=tiled_values,
-        repetitions=self._expectation_samples).to_tensor()
-    unique_expectations = tf.map_fn(
-        lambda x: tf.math.reduce_mean(observable.energy(x)),
-        samples,
-        fn_output_signature=tf.float32)
-    return utils.expand_unique_results(unique_expectations, idx)
+
+    @tf.custom_gradient
+    def _inner_expectation():
+      """Enables derivatives."""
+      unique_states, idx, _ = utils.unique_bitstrings_with_counts(
+          initial_states)
+      u = self.circuit + observable.circuit_dagger
+      unique_circuits = u(unique_states)
+      num_unique_circuits = tf.shape(unique_circuits)[0]
+      unique_tiled_values = tf.tile(
+          tf.expand_dims(u.symbol_values, 0), [num_unique_circuits, 1])
+      unique_samples = self._sample_layer(
+          unique_circuits,
+          symbol_names=u.symbol_names,
+          symbol_values=unique_tiled_values,
+          repetitions=self._expectation_samples).to_tensor()
+      with tf.GradientTape() as thetas_tape:
+        unique_expectations = tf.map_fn(
+            lambda x: tf.math.reduce_mean(observable.energy(x)),
+            unique_samples,
+            fn_output_signature=tf.float32)
+        forward_pass = tf.expand_dims(
+            utils.expand_unique_results(unique_expectations, idx), 1)
+
+      def grad_fn(*upstream, variables):
+        """Use `get_gradient_circuits` method to get QNN variable derivatives"""
+        thetas_gradients = thetas_tape.gradient(
```

Comment:
nit: would move this to the bottom of grad_fn since it sits unused for so many lines

Reply:
Done